File: area_scan_3d_camera/Advanced/CaptureStereo2DImages/CaptureStereo2DImages.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can obtain and save the stereo 2D images.
*/

#include <opencv2/highgui.hpp>
#include "area_scan_3d_camera/Camera.h"
#include "area_scan_3d_camera/api_util.h"

int main()
{
    mmind::eye::Camera camera;
    if (!findAndConnect(camera))
        return -1;

    mmind::eye::Frame2D stereoLeft, stereoRight;
    auto errorStatus = camera.captureStereo2D(stereoLeft, stereoRight, false);
    if (!errorStatus.isOK()) {
        showError(errorStatus);
        return -1;
    }

    cv::Mat cvMatLeft, cvMatRight;

    switch (stereoLeft.colorType()) {
    case mmind::eye::ColorTypeOf2DCamera::Monochrome:
    {
        mmind::eye::GrayScale2DImage grayImageLeft = stereoLeft.getGrayScaleImage();
        cvMatLeft =
            cv::Mat(grayImageLeft.height(), grayImageLeft.width(), CV_8UC1, grayImageLeft.data());

        mmind::eye::GrayScale2DImage grayImageRight = stereoRight.getGrayScaleImage();
        cvMatRight = cv::Mat(grayImageRight.height(), grayImageRight.width(), CV_8UC1,
                             grayImageRight.data());
        break;
    }
    case mmind::eye::ColorTypeOf2DCamera::Color:
    {
        mmind::eye::Color2DImage colorImageLeft = stereoLeft.getColorImage();
        cvMatLeft = cv::Mat(colorImageLeft.height(), colorImageLeft.width(), CV_8UC3,
                            colorImageLeft.data());

        mmind::eye::Color2DImage colorImageRight = stereoRight.getColorImage();
        cvMatRight = cv::Mat(colorImageRight.height(), colorImageRight.width(), CV_8UC3,
                             colorImageRight.data());
        break;
    }
    default:
        break;
    }

    const std::string imageFileLeft = "stereo2D_left.png";
    const std::string imageFileRight = "stereo2D_right.png";
    // cv::imshow(imageFileLeft, cvMatLeft);
    // cv::imshow(imageFileRight, cvMatRight);

    cv::imwrite(imageFileLeft, cvMatLeft);
    cv::imwrite(imageFileRight, cvMatRight);
    std::cout << "Capture and save the stereo 2D image: " << imageFileLeft << " and "
              << imageFileRight << std::endl;
    // cv::waitKey(0);

    camera.disconnect();
    std::cout << "Disconnected from the camera successfully." << std::endl;
    return 0;
}

File: area_scan_3d_camera/advanced/capture_stereo_2d_images.py
# With this sample, you can obtain and save the stereo 2D images.

import cv2

from mecheye.shared import *
from mecheye.area_scan_3d_camera import *
from mecheye.area_scan_3d_camera_utils import find_and_connect


class CaptureStereo2DImages(object):
    def __init__(self):
        self.camera = Camera()

    def capture_stereo_2d_images(self):
        stereo_left = Frame2D()
        stereo_right = Frame2D()
        error = self.camera.capture_stereo_2d(stereo_left, stereo_right)
        if not error.is_ok():
            show_error(error)
            return
        if stereo_left.color_type() == ColorTypeOf2DCamera_Monochrome:
            image_left = stereo_left.get_gray_scale_image()
            image_right = stereo_right.get_gray_scale_image()
        elif stereo_right.color_type() == ColorTypeOf2DCamera_Color:
            image_left = stereo_left.get_color_image()
            image_right = stereo_right.get_color_image()

        image_file_left = "stereo2D_left.png"
        image_file_right = "stereo2D_right.png"
        # cv2.imshow(image_file_left, image_left.data())
        # cv2.imshow(image_file_right, image_right.data())
        cv2.waitKey(0)

        cv2.imwrite(image_file_left, image_left.data())
        cv2.imwrite(image_file_right, image_right.data())
        print("Capture and save the stereo 2D images: {} and {}".format(
            image_file_left, image_file_right))

    def main(self):
        if find_and_connect(self.camera):
            self.capture_stereo_2d_images()
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")


if __name__ == '__main__':
    a = CaptureStereo2DImages()
    a.main()
